atef/widgets/config/find_replace.py

```diff
@@ -0,0 +1,1009 @@
+
+    def __init__(
+        self,
+        *args,
+        data: FindReplaceAction,
+        **kwargs
+    ) -> None:
+        super().__init__(*args, *kwargs)
+        self.data = data
+
+        last_dclass, attr = get_deepest_dataclass_in_path(self.data.path)
+        dclass_type = type(last_dclass).__name__
+        self.dclass_label.setText(f'{dclass_type}.{attr}')
+
+        pre_text = str(get_item_from_path(self.data.path, item=self.data.target))
+        # html wrapping to get some line wrapping
+        self.pre_label.setText(f'<font>{pre_text}</font>')
+        self.pre_label.setToolTip(f'<font>{pre_text}</font>')
+        preview_file = copy.deepcopy(self.data.target)
+
+        preview_success = data.apply(target=preview_file)
+        if preview_success:
+            post_text = str(get_item_from_path(self.data.path, item=preview_file))
+        else:
+            logger.warning('Unable to generate preview, provided replacement '
+                           'text is invalid')
+            post_text = '[INVALID]'
+        self.post_label.setText(f'<font>{post_text}</font>')
+        self.post_label.setToolTip(f'<font>{post_text}</font>')
+
+        ok_button = self.button_box.button(QtWidgets.QDialogButtonBox.Ok)
+        ok_button.setText('')
+        ok_button.setIcon(qta.icon('ei.check'))
+        delete_button = self.button_box.button(QtWidgets.QDialogButtonBox.Cancel)
+        delete_button.setText('')
+        delete_icon = self.style().standardIcon(
+            QtWidgets.QStyle.SP_TitleBarCloseButton
+        )
+        delete_button.setIcon(delete_icon)
+
+        # path details
+        path_list = []
+        for segment in self.data.path:
+            if isinstance(segment[0], str):
+                name = segment[0]
+            else:
+                name = type(segment[0]).__name__
+            path_list.append(f'({name}, {segment[1]})')
+
+        path_str = '>'.join(path_list)
+        detail_widget = QtWidgets.QLabel(path_str + '\n')
+        detail_widget.setWordWrap(True)
+
+        widget_action = QtWidgets.QWidgetAction(self.details_button)
+        widget_action.setDefaultWidget(detail_widget)
+
+        widget_menu = QtWidgets.QMenu(self.details_button)
+        widget_menu.addAction(widget_action)
+        self.details_button.setMenu(widget_menu)
+
+        self.button_box.accepted.connect(self.apply_action)
+        self.button_box.rejected.connect(self.reject_action)
+
+    def apply_action(self) -> None:
+        success = self.data.apply()
+        if not success:
+            QtWidgets.QMessageBox.warning(
+                self,
+                'Edit was not applied successfully, and will be removed'
+            )
+        self.remove_item.emit()
+
+    def reject_action(self) -> None:
+        self.remove_item.emit()
+
+
+class FillTemplatePage(DesignerDisplay, QtWidgets.QWidget):
+
+    file_name_label: QtWidgets.QLabel
+    type_label: QtWidgets.QLabel
+
+    device_table: QtWidgets.QTableWidget
```

Comment:
I think the biggest feature you could add is the ability to change (add/remove/edit) devices and signals directly in this table using the device/component selector dialog:
<img width="343" alt="image" src="https://github.com/pcdshub/atef/assets/5139267/53c01d81-1f49-402d-8e21-6faef0c5bd62">

The find/replace stuff is very valuable, but I think the primary thing people will want to template is "test that device instead of this device"

Reply:
aaaah now that's a fancy idea.  I'll make a followup, because my vision of this quickly became very complicated